### PR TITLE
=str Skip parsing when buffer size < 1

### PR DESCRIPTION
**Pull request by He-Pin**
I think we can do a quic returning if the buffer size is zero.

### CODE REVIEW

**Review by mdedetrich — CHANGES_REQUESTED** [review]
This might be considered excessive but considering we have had a regression in this area of Pekko, would it be possible to add a unit test which checks that when the buffer size is 0 to check that both current Pekko and the change in this PR both return `false`?

The easiest way to do this is to make `seekObject()` package private (so you can access it in a test) and construct `JsonObjectParser` (which initialises `buffer` to empty so `buffer.length` is zero) and immediately call `seekObject()` to see if it returns `false`.

If this is too difficult then let me know and I can reconsider.

**Comment by He-Pin** [comment]
It can be false if the downstream poll, but the upstream is not ready, or downstream is faster and upstream is idle.

I can add more detailed test for JsonFramming but I don't see much value to add a test for the short-circuit logic，as  it's identical to the objectCompleted.

As the chances this short-circuit can benefit, i think that maybe why it's not there in the firsr place.

**Comment by mdedetrich** [comment]
tbh I am not entirely familiar with this section of Pekko, I am just being cautious/skeptical because we had a general regression here which I am trying to avoid in the future.

If you are happy with adding a more detailed `JsonFraming` test that would be great, I suspect you would do this in a separate PR? If so let me know and can review this in more detail in the next couple of days.

**Comment by pjfanning** [comment]
Could we do this in v1.0.1? It's not a bug and it just doesn't feel urgent.

**Comment by mdedetrich** [comment]
Agreed

**Comment by He-Pin** [comment]
let's defer this.

**Comment by mdedetrich** [comment]
> i want to add some random tests against JsonFraming.

Perfect thanks! Such tests should definitely be merged before 1.0.0 if they are ready in time to give us extra confidence.

**Comment by mdedetrich** [comment]
@He-Pin I am adding this to 1.1.x milestone

File: stream/src/main/scala/org/apache/pekko/stream/impl/JsonObjectParser.scala [hunk 1]
```diff
@@ -115,7 +115,9 @@ import pekko.util.ByteString
   private def seekObject(): Boolean = {
     completedObject = false
     val bufSize = buffer.length
-
+    if (bufSize == 0) {
```
Comment (pjfanning) [inline stream/src/main/scala/org/apache/pekko/stream/impl/JsonObjectParser.scala:118]: I don't like 'return' statements. Can't you just do this ?

```
if (bufSize > 0) {
  // add everything from the rest if the function here
}
```

Comment (jrudolph) [reply stream/src/main/scala/org/apache/pekko/stream/impl/JsonObjectParser.scala:118]: Agreed, maybe not super important in that case, but let's avoid a potential exception being thrown.

**Review by pjfanning — APPROVED** [review]
lgtm

File: stream/src/main/scala/org/apache/pekko/stream/impl/JsonObjectParser.scala [hunk 2]
```diff
@@ -115,20 +115,21 @@ import pekko.util.ByteString
   private def seekObject(): Boolean = {
     completedObject = false
     val bufSize = buffer.length
+    if (bufSize > 1) {
```
Comment (He-Pin) [inline stream/src/main/scala/org/apache/pekko/stream/impl/JsonObjectParser.scala:118]: a `{}` is minimal 2 bytes.

**Review by mdedetrich — APPROVED** [review]
lgtm

**Comment by He-Pin** [comment]
The error is unrelated.

**Pull request merged**